### PR TITLE
[9.x] Support where and orderby over relation aggregations

### PR DESCRIPTION
**Pull request by khalilst**
This PR adds the `where` and `orderBy` clauses over relationships.

---

Examples of `where` clause (assume comments table has a `likes` column):

```php
Post::whereHasCount('comments', '>', 10);
Post::whereHasMax('comments', 'likes', '<=', 5);
Post::whereHasMin('comments', 'likes', '>=', 100);
Post::whereHasSum('comments', 'likes', '>', '1000');
Post::whereHasAvg('comments', 'likes', '>', 1000);
```
They all are using `whereHasAggregate()` method.
This method is able to be used with other functions like `distinct` or custom functions.

---

Examples of `orderBy` clause:

```php
Post::orderByCount('comments');
Post::orderByMax('comments', 'likes', 'desc');
Post::orderByMin('comments', 'likes');
Post::orderBySum('comments', 'likes');
Post::orderByAvg('comments', 'likes');
```

Similar to the `where` methods, the `orderBy` methods are using `orderByAggregate()` method.
And it is able to be used with other functions like `distinct` or custom functions.

---

### Implementation Notes:
At a higher level, these methods are using the same logic from `withAggregate()` method and in order to avoid duplicating code another method was created to have that logic [`relationAggregate()`]. 
In summary we have this hierarchy:

```mermaid
flowchart TB
    subgraph withAggregate
        withCount
        withMax
        withMin
        withSum
        withAvg
        withExists
    end
    subgraph whereHasAggregate
        whereHasCount
        whereHasMax
        whereHasMin
        whereHasSum
        whereHasAvg
    end
    subgraph orderByAggregate
        orderByCount
        orderByMax
        orderByMin
        orderBySum
        orderByAvg
    end
    relationAggregate-->withAggregate;
    relationAggregate-->whereHasAggregate;
    relationAggregate-->orderByAggregate;
```

### CODE REVIEW

**Comment by ankurk91** [comment]
no tests?

**Comment by khalilst** [comment]
@ankurk91 The core changes are already covered in tests for `withCount()` and `withMin()`.
If the idea was accepted at first glance, I would add tests as well.

**Comment by khalilst** [comment]
Test cases implemented: 

- `whereHasAggregation() Family`: 4b740c6538ceb440aa086fdcbd638bf73c978adb
- `orderByAggregation() family`: b926af3b77e5581db89bd07fa22ca74eec095b9a
- Fix for failing test related to `RelationCountHash` regex: 7101b9ed0a1a913fe658ca5523ca53aa69f6d6ee

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.
